apps/src/templates/rubrics/LearningGoals.jsx

```diff
@@ -28,6 +29,149 @@ import ProgressRing from './ProgressRing';
+          if (index < 0) {
+            // Remember the first position in the original code that
+            // the AI references.
+            index = position;
+          }
+          // Remember the last position in the original code.
+          lastIndex = position + lineLength;
+          context = context.substring(line.length + 1).trim();
+          if (context === '') {
+            // All of the code was found
+            break;
+          }
+        } else {
+          // We didn't match it. Reset our search.
+          context = snippet;
+          index = -1;
+        }
+      }
+
+      // Move the current search position
+      position += lineLength + 1;
+    }
+  }
+
+  // As long as we found a region, we will determine what the line
+  // number was for the first and last line of the region.
+  if (index < 0) {
+    return [null, null];
+  }
+
+  let lineNumber = (code.substring(0, index).match(/\n/g) || []).length + 1;
+  let lastLineNumber =
+    (code.substring(0, lastIndex).match(/\n/g) || []).length + 1;
+
+  return [lineNumber, lastLineNumber];
+}
+
+/**
+ * Adds annotations to the source code viewed in the current editor based
+ * on the AI observations of the learning goal referenced by the given index.
+ */
+export function annotateLines(observations) {
+  // Get a reference to all the whitespaced lines of code
+  const code = getAnonymizedCode();
+  const lines = code.split('\n');
+
+  // Go through the AI observations
+  // For every reference the AI gave us, we will find it in the code.
+  // The AI has trouble giving line numbers, so even though we parse
+  // those out, we do not trust them and instead find the code it
+  // references to highlight it.
+  for (const match of observations.matchAll(
+    'Line (\\d+)(?:\\s*-\\s*(\\d+))?:(.+?)\\s*(?=Line|$)'
+  )) {
+    let lineNumber = parseInt(match[1]);
+    let lastLineNumber = parseInt(match[2] || lineNumber);
+    let found = false;
+    const context = match[3].trim();
+
+    const message = context
+      .substring(0, context.indexOf('`') || context.length)
+      .trim();
+
+    // We look at all of the code references the AI gave us which are
+    // surrounded by backticks.
+    const references = context.substring(message.length);
+    for (const submatch of references.matchAll(/`([^\`]+)`/g)) {
+      let snippet = submatch[1].trim();
+
+      let [lineNumber, lastLineNumber] = findCodeRegion(code, lines, snippet);
+
+      // Annotate that first line and highlight all lines, if they were found
+      if (lineNumber && lastLineNumber) {
+        found = true;
+        studioApp().annotateLine(message, lineNumber);
+        for (let i = lineNumber; i <= lastLineNumber; i++) {
+          studioApp().highlightLine(i);
+        }
+      }
+    }
+
+    if (!found) {
+      studioApp().annotateLine(message, lineNumber);
+      for (let i = lineNumber; i <= lastLineNumber; i++) {
+        studioApp().highlightLine(i);
```

Comment:
How (un)reliable are line numbers provided by AI right now? I'm wondering if we should try to note that the line number is approximate within the message. one idea would be to include the original text generated by AI:
```
Line 1: elusive code: 'var foo;'
```
both so that we can see that we were unable to find the line number, and also so that the teacher may be able to infer which line the AI was referring to in the case that it was not far off. 

this can just be an idea for later if it sounds like too much for now

Reply:
Ah, that's a good point. Surprisingly, it doesn't seem to generate code in the responses that I cannot find... but it makes sense to include all of the context when it cannot determine where the annotation should actually go.